### DISCUSSION TITLE
Data in pytest tests

### DISCUSSION THREAD

**Discussion by davidorme** [ImperialCollegeLondon/virtual_rainforest#77] [Q&A] [answered]
I'm starting to play around with data loading code, and that involves providing data to load in tests.

One option is provide data in the test files themselves - and for simple text/JSON objects that is probably easiest. However, for more complex files such as NetCDF, creating `ByteIO` objects within test files so the data only ever exists in memory just seems like it adds a lot of complex code to the test file and having the actual files seems easier?

I've had a look at a couple of options:

* [pytest-datafiles](https://pypi.org/project/pytest-datafiles/)
* [pytest-datadir](https://pypi.org/project/pytest-datadir/)

Both of those are currently 2.7 and then 3.8 or lower, but `pytest-datadir` - which I think has the better API -  looks to be on the verge of a new release that drops 2.7 and goes up to 3.10. As in, we could probably add a poetry requirement of their `master` branch as a stop gap right now.

But - any other options or thoughts?

**Comment by dalonsoa** [comment, marked as answer]
I'm not sure I understand the issue here. Is there any problem on having real files for tests, JSON, NetCDF or anything else? Unless they have sensitive data or are huge files (>50 Mb), that's what we have done in other projects, just putting them within a subfolder `data` inside `tests`, and create an associated feature in `conftest.py` so they can be used easily within any test.

**Reply by davidorme** [reply]
I may be over-engineering again! I've seen comments about making sure the tests work on independent copies of files. The `python-datadirs` package provides a couple of built-in fixture that essentially removes the need to adding a feature in `conftest.py`. 

What kind of feature do you add to `conftest.py`? Do you mean create a fixture that points to particular sets of files?

**Reply by dalonsoa** [reply]
It depends of what you need. It might be just the path to the real file. It might be the path to a temporal copy to the file in a temp directory, so it is always a fresh file with every use (this might be what that package does - I've never used it). It might be the file already open and provided as a stream... Up to you, to be honest. But all cases start with a real file in the directory tree.

**Reply by davidorme** [reply]
OK - in which case, I think `python-datadirs` is just a plugin that automatically create fixtures that handle that (and uses temporary files automatically). Which seems useful - I suspect a lot of tests may require data, so having those fixtures defined for us seems attractive.

**Reply by dalonsoa** [reply]
That sounds ok, but just a note of caution here: the more dependencies yo have , the higher the chance for they to break at some point. If the only advantage is to save you creating a simple fixture without adding anything else to the table, it might not be worth it.

**Reply by davidorme** [reply]
I completely agree that we want to keep dependencies to a minimum. I'll hang on to it for the moment, but it is at least easy to reverse - not a complex API that we'd need to reinvent, simply a shortcut for what may be a common use case.
